### ISSUE TITLE
APF: Understand the failures of 5k scalability tests with client-side rate-limitting disabled

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#115293]
The results can be found in:
https://github.com/kubernetes/kubernetes/pull/109614

Facts:
- latency metrics include waiting time in APF
- so basically, long waiting time (saturated PLs) would result in exactly this

Hypothesis:
- some PLs are saturated and thus latency metrics are exceeding the threshold

But we need to confirm this and figure out how to improve it.

[The next step would be to see how we can speed up this test even further and see what will happen later, but I think we're currently mostly blocked on scheduler throughput.]

/assign @MikeSpreitzer 
/cc @tkashem

**Comment by MikeSpreitzer** [comment]
So I started looking at https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/109614/pull-kubernetes-e2e-gce-scale-performance-manual/1612926506110029824, which complained

{ Failure :0
[measurement call APIResponsivenessPrometheus - APIResponsivenessPrometheusSimple error: top latency metric: there should be no high-latency requests, but: [got: &{Resource:pods Subresource: Verb:DELETE Scope:resource Latency:perc50: 106.194648ms, perc90: 1.112666683s, perc99: 2.292180059s Count:553900 SlowCount:65905}; expected perc99 <= 1s]]
:0}

And indeed I can see that also in the APF metrics for queuing delay.  The big loser in the following picture is the `kube-system-service-accounts` flow schema, which goes in the `workload-high` priority level.  The next biggest loser is the `system-nodes` flow schema, which goes in the `system` priority level.

<img width="932" alt="Screen Shot 2023-01-31 at 2 31 08 AM" src="https://user-images.githubusercontent.com/14296719/215699808-807f7e8a-7dc5-4501-82ec-46a2724307ce.png">

The next graph shows the dynamic concurrency limits.  As the `workload-high` priority level got overloaded, it borrowed a lot.

<img width="926" alt="Screen Shot 2023-01-31 at 2 42 32 AM" src="https://user-images.githubusercontent.com/14296719/215700777-9c48350d-7f8b-4b55-97c1-597a0b0031aa.png">

The next picture shows seat usage/limit.  Interestingly, the `workload-high` priority level started occupying all of its seats shortly _before_ 23:00 (UTC; sorry, some of the graphs are in EST).  But the queuing delay did not shoot up until about 23:05, and also at that time the `system` priority level also started using all of its seats.

<img width="926" alt="Screen Shot 2023-01-31 at 2 45 58 AM" src="https://user-images.githubusercontent.com/14296719/215701334-a7d318ea-1cb4-48cd-9690-b50786075d55.png">

**Comment by MikeSpreitzer** [comment]
BTW, here is a graph of the SmoothSeatDemand values used in the borrowing logic (https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1040-priority-and-fairness#dispatching).

<img width="925" alt="Screen Shot 2023-01-31 at 2 33 15 AM" src="https://user-images.githubusercontent.com/14296719/215703211-eda7520a-8191-4aa0-9594-3c8ec5356f52.png">

**Comment by wojtek-t** [comment]
To give timeline of the test that seems to be important:
```
I0110 22:57:48.173918   90570 simple_test_executor.go:149] Step "[step: 21] scale and update objects" started
I0110 23:00:35.320727   90570 simple_test_executor.go:171] Step "[step: 21] scale and update objects" ended
I0110 23:00:35.320782   90570 simple_test_executor.go:149] Step "[step: 22] Waiting for 'scale and update objects' to be completed" started
I0110 23:05:00.339521   90570 simple_test_executor.go:171] Step "[step: 22] Waiting for 'scale and update objects' to be completed" ended
I0110 23:05:00.339639   90570 simple_test_executor.go:149] Step "[step: 23] Starting measurement for 'delete objects'" started
...`

So to your point @MikeSpreitzer 

- before 23:00 to 23:05 - there was "scale" phase of the test
- at 23:05, the "delete" phase started

It probably explains why the load right before 23:05 is visibly different than the one right after 23:05.
It seems that "scale" phase is expensive, but not overloading the cluster.
The "delete" phase seems to be visibly more expensive.

**Comment by MikeSpreitzer** [comment]
is the scale phase driven by the client with its own pacing, and the delete phase driven by a controller without pacing?

**Comment by MikeSpreitzer** [comment]
The test is defined in https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-scalability/sig-scalability-presubmit-jobs.yaml

**Comment by MikeSpreitzer** [comment]
And that test references the following.

1. https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/testing/load/config.yaml
2. https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/testing/huge-service/config.yaml
3. https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/testing/access-tokens/config.yaml

**Comment by MikeSpreitzer** [comment]
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/testing/load/config.yaml invokes https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/testing/load/modules/reconcile-objects.yaml three times: first with `actionName` "create", second with `actionName` "scale and update", and finally with `actionName` "delete".

Following are the timings from that run.

I0110 22:27:53.576504   90570 simple_test_executor.go:149] Step "[step: 06] create objects" started
I0110 22:30:38.129830   90570 simple_test_executor.go:171] Step "[step: 06] create objects" ended
I0110 22:54:17.874831   90570 simple_test_executor.go:171] Step "[step: 07] Waiting for 'create objects' to be completed" ended

I0110 22:57:48.173918   90570 simple_test_executor.go:149] Step "[step: 21] scale and update objects" started
I0110 23:00:35.320727   90570 simple_test_executor.go:171] Step "[step: 21] scale and update objects" ended
I0110 23:05:00.339521   90570 simple_test_executor.go:171] Step "[step: 22] Waiting for 'scale and update objects' to be completed" ended

I0110 23:05:00.339773   90570 simple_test_executor.go:149] Step "[step: 24] delete objects" started
I0110 23:08:06.027767   90570 simple_test_executor.go:171] Step "[step: 24] delete objects" ended
I0110 23:17:56.526108   90570 simple_test_executor.go:171] Step "[step: 25] Waiting for 'delete objects' to be completed" ended

**Comment by wojtek-t** [comment]
> is the scale phase driven by the client with its own pacing, and the delete phase driven by a controller without pacing?

It shouldn't generally, but there is a caveat here (which is also why we don't see any problems in the create phase):
- even though we create pods quickly, scheduler actually has a limited throughput. Given that the main load is triggered by scheduled pods (scheduled pods triggers its starting and later updating corresponding EndpointsSlice which is expensive as watched by all nodes), we get this somewhat rate-limitted.
- in update phase, this is randomize, so roughly half of pod changes are creations and half are deletions - so we're still somewhat rate-limitted by scheduler
- this is not anymore the case in delete latency

So basically what we can try is to disable some of scheduler's functionality (as a test) to improve its throughput and see how that will influence the load on the kube-apiserver and the test itself.

I will try to update my test PR to do that.

**Comment by MikeSpreitzer** [comment]
Here is utilization of the 64 "CPUs".  It was fairly busy throughout the creates, the scale and updates, and the deletes.

<img width="1004" alt="Screen Shot 2023-02-06 at 1 17 47 PM" src="https://user-images.githubusercontent.com/14296719/217053872-c4f09109-771d-4b49-a204-aaf77e78bb7b.png">

**Comment by MikeSpreitzer** [comment]
Here are the kube API operation rates on pods.  There is a lot going on here.

The big green spike around 22:30 is the creates.
The lower and slower responses are:
highest - GET of whole pod and PATCH of status, visually nearly identical;
lower - POST of binding.

During the scale and updates, in decreasing order of max height:
purple - GET of whole pod, 200 response code;
purple - DELETE of whole pod;
aqua - PATCH of status;
mauve - GET of whole pod, 404 response code;
gray - POST of binding.

During the deletes, in decreasing order of max height:
purple - GET of pod, 200 response code;
purple - DELETE of pod, 200 response code;
aqua - PATCH of status;
blue - DELETE of pod, 429 response code;
purple - GET of pod, 404 response code.

<img width="983" alt="Screen Shot 2023-02-06 at 1 24 29 PM" src="https://user-images.githubusercontent.com/14296719/217054326-ab9d867d-30d5-413e-8e29-5b747bf3791d.png">

**Comment by MikeSpreitzer** [comment]
Here is the average time to execute a request (inside APF), excluding the peak after 23:30.

The highest peak is blue, FS=kube-system-service-accounts PL=workload-high, during deletes at 23:07:54.  This is also the blue peak at 23:13:20.

The next highest peak is gray, FS=workload-leader-election PL=leader-election, during scale and updates at 22:59:43.

Next down is also gray, FS=system-nodes PL=system, during scale and updates at 22:58:27.

During the creates, the first and highest peak is FS=kube-system-service-acounts PL=workload-high at 22:28:03.  Nearly as high and almost impossible to visually distinguish is FS=endpoint-controller PL=workload-high.  Also in the visual mix is FS=kube-scheduler PL=workload-high.

The green peaking at 23:34:42 is FS=probes PL=exempt.

The gray peaking at 22:45:09 is FS=workload-leader-election PL=leader-election.

<img width="986" alt="Screen Shot 2023-02-06 at 1 38 18 PM" src="https://user-images.githubusercontent.com/14296719/217056995-5e0d2f45-7a8e-4deb-b6b2-85f023a04a7c.png">

**Comment by MikeSpreitzer** [comment]
Here is a Grafana panel focusing on the five flow schemas called out in the previous comment.

<img width="1256" alt="Screen Shot 2023-02-06 at 2 23 06 PM" src="https://user-images.githubusercontent.com/14296719/217065542-6341cda1-1805-4d2b-b12a-4eb55f476aa5.png">

**Comment by MikeSpreitzer** [comment]
Here is where the apiserver was spending its etcd time.

<img width="1249" alt="Screen Shot 2023-02-06 at 2 43 59 PM" src="https://user-images.githubusercontent.com/14296719/217069464-038fdd3d-2a5f-416c-b6ae-b77aae99fdf4.png">

**Comment by MikeSpreitzer** [comment]
In comparison, https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/113263/pull-kubernetes-e2e-gce-scale-performance-manual/1622546468801875968 was rated a success.  Following are the timings of the three uses of reconcile-objects.

I0206 11:41:09.548158   91661 simple_test_executor.go:149] Step "[step: 06] create objects" started
I0206 11:43:53.835473   91661 simple_test_executor.go:171] Step "[step: 06] create objects" ended
I0206 12:09:29.008236   91661 simple_test_executor.go:171] Step "[step: 07] Waiting for 'create objects' to be completed" ended

I0206 12:15:56.456550   91661 simple_test_executor.go:149] Step "[step: 21] scale and update objects" started
I0206 12:18:45.478611   91661 simple_test_executor.go:171] Step "[step: 21] scale and update objects" ended
I0206 12:27:30.487530   91661 simple_test_executor.go:171] Step "[step: 22] Waiting for 'scale and update objects' to be completed" ended

I0206 12:27:30.487672   91661 simple_test_executor.go:149] Step "[step: 24] delete objects" started
I0206 12:30:36.413458   91661 simple_test_executor.go:171] Step "[step: 24] delete objects" ended
I0206 13:02:21.918803   91661 simple_test_executor.go:171] Step "[step: 25] Waiting for 'delete objects' to be completed" ended

**Comment by MikeSpreitzer** [comment]
As another point of comparison, https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/109614/pull-kubernetes-e2e-gce-scale-performance-manual/1622499514118049792 as another failure of #109614 .  Following are the timings of its uses of reconcile-objects.

I0206 08:28:54.889940   91537 simple_test_executor.go:149] Step "[step: 06] create objects" started
I0206 08:31:39.272597   91537 simple_test_executor.go:171] Step "[step: 06] create objects" ended
I0206 08:55:14.532275   91537 simple_test_executor.go:171] Step "[step: 07] Waiting for 'create objects' to be completed" ended

I0206 08:58:43.207482   91537 simple_test_executor.go:149] Step "[step: 21] scale and update objects" started
I0206 09:01:31.050857   91537 simple_test_executor.go:171] Step "[step: 21] scale and update objects" ended
I0206 09:06:01.330872   91537 simple_test_executor.go:171] Step "[step: 22] Waiting for 'scale and update objects' to be completed" ended

I0206 09:06:01.331121   91537 simple_test_executor.go:149] Step "[step: 24] delete objects" started
I0206 09:09:06.454109   91537 simple_test_executor.go:171] Step "[step: 24] delete objects" ended
I0206 09:18:50.845408   91537 simple_test_executor.go:171] Step "[step: 25] Waiting for 'delete objects' to be completed" ended

**Comment by MikeSpreitzer** [comment]
The timings of steps 6, 21, and 24 are driven by the client using a golang library.
The completion times are affected by this experiment.

j10 bad completion: 23:40, 4:25, 9:50
f6 good completion: 25:36, 8:45, 31:45
f6 bad completion: 23:35, 4:30, 9:44

The change in #109614 did not much speed up the creates, and greatly sped up the deletes.

**Comment by MikeSpreitzer** [comment]
Here is a graph of the CPU utilization in the Feb 6 run with client-side rate-limiting disabled.

<img width="996" alt="Screen Shot 2023-02-06 at 3 57 55 PM" src="https://user-images.githubusercontent.com/14296719/217084072-9740145c-3d73-4706-bfd2-c927b187768f.png">

**Comment by wojtek-t** [comment]
This test is still failing.
But I think the main and first problem that we should solve here is to fix: https://github.com/kubernetes/kubernetes/issues/104809

[I think that requests going to exempt PL may be the ones that are overloading us]

**Comment by Vyom-Yadav** [comment]
Hey @MikeSpreitzer @wojtek-t 
This test is still flaking https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-scalability-100.
- [2024-06-26 14:40:15 +0000 UTC](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/1805974172216594432)
- [2024-06-26 10:52:15 +0000 UTC](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/1805916793835884544)
- [2024-06-26 03:17:16 +0000 UTC](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/1805802286744604672)
